enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,358 @@
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+We propose that when utilizing Azure Workload Identity, the credentials secret will contain an `azure_client_id` that is the `clientID` of the Managed Identity provisioned by `ccoctl` for the operator. The `azure_client_secret` key will be absent and instead we can provide the path to the mounted `ServiceAccount` token as an `azure_federated_token_file` key; the path to the mounted token is well
+known and is specified in the operator deployment.
+
+The resource group in which the installer will create infrastructure will not be known when these secrets are generated by `ccoctl` ahead of installation and operators which rely on `azure_resourcegroup` and `azure_resource_prefix` such as the
+[image-registry](https://github.com/openshift/cluster-image-registry-operator/blob/8556fd48027f89e19daad36e280b60eb93d012d4/pkg/storage/azure/azure.go#L95-L100) should obtain the resource group details from the cluster `Infrastructure` object instead.
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_federated_token_file: <path to mounted service account token, eg. "/var/run/secrets/openshift/serviceaccount/token">
+  azure_region: <region>
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+#### Creating workload identity clients in operators
+
+In order to create Azure clients which utilize a `ClientAssertionCredential`, operators must update to version `>= v1.3.0-beta.4` of the azidentity package within [azure-sdk-for-go](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.3.0-beta.4). Ahead of this work, due to the [end of life
+announcement](https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/microsoft-entra-change-announcements-september-2022-train/ba-p/2967454) of the Azure Active Directory Authentication Library (ADAL), PRs (ex. [openshift/cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/pull/846)) have been opened for operators to migrate to creating clients via
+azidentity which are converted into an authorizer for use with v1 clients. Once these changes have been made, we propose that OpenShift operators continue to utilize a config secret to obtain authentication details as described in the previous section but create workload identity clients when the `azure_client_secret` is absent and when  `azure_federated_token_file` fields are found in the
+config. Config secrets will be generated by cluster creators prior to installation by using `ccoctl` and will be provided as manifests for install.
+
+Due to the deployment of the Azure Workload Identity mutating admission webhook, environment variables should also be respected by client instantiation as an alternative way of supplying the `clientID` eg. `AZURE_CLIENT_ID`, `tenantID` eg. `AZURE_TENANT_ID` and `federatedTokenFile` eg. `AZURE_FEDERATED_TOKEN_FILE`.
```

Comment:
why not simply use the azure recommendation and discontinue use of the secret?

Reply:
Previous discussion here https://github.com/openshift/enhancements/pull/1301#discussion_r1135622157 but our challenge would be ensuring that environment variables are set correctly for service principal auth vs. workload identity auth and we have plumbing for authenticating with details in the config secret today.

We couldn't use Azure's mutating webhook for service principal (client secret) based auth so the EP doesn't rely on the webhook for OpenShift operators for this reason. Additionally, relying on the webhook would mean that we need to annotate service accounts with the client information that we're putting in the secrets and I think this would require some ongoing reconciliation from CCO which is currently disabled on clusters in manual mode.

Am I overlooking that we could set ALL keys from the secret on all operators and allow the SDK to choose the auth method based? Too wide of a blast?

Reply:
Changing operators to use environment variables has a much larger scope; operators would either need to template their deployment to set environment variables based on the contents of our config secret OR some external process would have to configure the environment variables for the operator's deployment (for example: CCO controller + webhook). Also, operators themselves need to be rewritten to not consume the secret for the current client secret auth method.

Reply:
It's frustrating to see CVO limitations driving choices to deviate from methods that will work with existing auditing tools, but I'm willing to yield to practicality on this point to get started.